src/display/editor/outliner.js

```diff
@@ -317,6 +343,452 @@ class HighlightOutline extends Outline {
+  // The next 6 elements are the last 3 points of the bottom part of the
+  // outline.
+  // We track the last 3 points in order to be able to:
+  //  - compute the normal of the line,
+  //  - compute the control points of the quadratic Bézier curve.
+  #last = new Float64Array(18);
+
+  #min;
+
+  #min_dist;
+
+  #scaleFactor;
+
+  #thickness;
+
+  #points = [];
+
+  static #MIN_DIST = 8;
+
+  static #MIN_DIFF = 2;
+
+  static #MIN = FreeOutliner.#MIN_DIST + FreeOutliner.#MIN_DIFF;
+
+  constructor({ x, y }, box, scaleFactor, thickness, innerMargin = 0) {
+    this.#box = box;
+    this.#thickness = thickness * scaleFactor;
+    this.#last.set([NaN, NaN, NaN, NaN, x, y], 6);
+    this.#innerMargin = innerMargin;
+    this.#min_dist = FreeOutliner.#MIN_DIST * scaleFactor;
+    this.#min = FreeOutliner.#MIN * scaleFactor;
+    this.#scaleFactor = scaleFactor;
+    this.#points.push(x, y);
+  }
+
+  get free() {
+    return true;
+  }
+
+  isEmpty() {
+    return isNaN(this.#last[8]);
+  }
+
+  add({ x, y }) {
+    const [layerX, layerY, layerWidth, layerHeight] = this.#box;
+    let [x1, y1, x2, y2] = this.#last.subarray(8, 12);
+    const diffX = x - x2;
+    const diffY = y - y2;
+    const d = Math.hypot(diffX, diffY);
+    if (d < this.#min) {
+      // The idea is to avoid garbage points around the last point.
+      // When the points are too close, it just leads to bad normal vectors and
+      // control points.
+      return false;
+    }
+    const diffD = d - this.#min_dist;
+    const K = diffD / d;
+    const shiftX = K * diffX;
+    const shiftY = K * diffY;
+
+    // We update the last 3 points of the line.
+    let x0 = x1;
+    let y0 = y1;
+    x1 = x2;
+    y1 = y2;
+    x2 += shiftX;
+    y2 += shiftY;
+
+    // We keep track of the points in order to be able to compute the focus
+    // outline.
+    this.#points?.push(x, y);
+
+    // Create the normal unit vector.
+    // |(shiftX, shiftY)| = |K| * |(diffX, diffY)| = |K| * d = diffD.
+    const nX = -shiftY / diffD;
+    const nY = shiftX / diffD;
+    const thX = nX * this.#thickness;
+    const thY = nY * this.#thickness;
+    this.#last.set(this.#last.subarray(2, 8), 0);
+    this.#last.set([x2 + thX, y2 + thY], 4);
+    this.#last.set(this.#last.subarray(14, 18), 12);
+    this.#last.set([x2 - thX, y2 - thY], 16);
+
+    if (isNaN(this.#last[6])) {
+      if (this.#top.length === 0) {
+        this.#last.set([x1 + thX, y1 + thY], 2);
+        this.#top.push(
+          NaN,
+          NaN,
+          NaN,
+          NaN,
+          (x1 + thX - layerX) / layerWidth,
+          (y1 + thY - layerY) / layerHeight
+        );
+        this.#last.set([x1 - thX, y1 - thY], 14);
+        this.#bottom.push(
+          NaN,
+          NaN,
+          NaN,
+          NaN,
+          (x1 - thX - layerX) / layerWidth,
+          (y1 - thY - layerY) / layerHeight
+        );
+      }
+      this.#last.set([x0, y0, x1, y1, x2, y2], 6);
+      return !isNaN(this.#last[8]);
```

Comment:
Should we use `isEmpty` here instead, since that seems a little clearer?

*Please note:* This question applies in multiple spots below as well.